src/languages/r.js

```diff
@@ -131,48 +146,91 @@ export default function(hljs) {
+      // ensure the number is not part of an identifier, and we cannot use
+      // negative look-behind assertions. So instead we explicitly handle all
+      // possible combinations of (operator|punctuation), number.
+      // TODO: replace with negative look-behind when available
+      // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+\.[0-9a-fA-F]*[pP][+-]?\d+i?/ },
+      // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+([pP][+-]?\d+)?[Li]?/ },
+      // { begin: /(?<![a-zA-Z0-9._])(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?[Li]?/ }
       {
         relevance: 0,
-        className: {
-          2: "number"
-        },
         variants: [
-          // TODO: replace with negative look-behind when available
-          // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+\.[0-9a-fA-F]*[pP][+-]?\d+i?/ },
-          // { begin: /(?<![a-zA-Z0-9._])0[xX][0-9a-fA-F]+([pP][+-]?\d+)?[Li]?/ },
-          // { begin: /(?<![a-zA-Z0-9._])(\d+(\.\d*)?|\.\d+)([eE][+-]?\d+)?[Li]?/ }
           {
-            // Special case: only hexadecimal binary powers can contain fractions.
+            scope: {
+              1: 'operator',
+              2: 'number'
+            },
             match: [
-              /[^a-zA-Z0-9._]/, // not part of an identifier
-              /0[xX][0-9a-fA-F]+\.[0-9a-fA-F]*[pP][+-]?\d+i?/
+              OPERATORS_RE,
+              NUMBER_TYPES_RE
             ]
           },
           {
+            scope: {
+              1: 'operator',
+              2: 'number'
+            },
             match: [
-              /[^a-zA-Z0-9._]/, // not part of an identifier
-              /0[xX][0-9a-fA-F]+(?:[pP][+-]?\d+)?[Li]?/
+              /%[^%]*%/,
+              NUMBER_TYPES_RE
             ]
           },
           {
+            scope: {
+              1: 'punctuation',
+              2: 'number'
+            },
             match: [
-              /[^a-zA-Z0-9._]/, // not part of an identifier
-              /(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?[Li]?/
+              PUNCTUATION_RE,
+              NUMBER_TYPES_RE
+            ]
+          },
+          {
+            scope: { 2: 'number' },
+            relevance: 0,
+            match: [
+              /[^a-zA-Z0-9._]|^/, // not part of an identifier, or start of document
+              NUMBER_TYPES_RE
             ]
           }
         ]
       },
+
+      // Operators/punctuation when they're not directly followed by numbers
       {
-        // infix operator
-        begin: '%',
-        end: '%'
+        // Relevance boost for the most common assignment form.
+        scope: { 3: 'operator' },
+        match: [
+          IDENT_RE,
+          /\s+/,
+          /<-/,
+          /\s+/
+        ]
       },
-      // relevance boost for assignment
+
+      {
+        scope: 'operator',
+        variants: [
+          {
+            relevance: 0,
+            match: OPERATORS_RE
+          },
+          { match: /%[^%]*%/ }
```

Comment:
> Any %…% sequence is a valid binary operator, %% included (and this also includes whitespace! So a % foo % b is a valid expression with two operands, a and b).

So I'm worried about %%... I've had similar issues with `||` in the past where a Ruby rule for `|param|` was eating ors (`||`) in other languages because it was coded too widely.  I wonder/worry about comments/patterns in some other languages.  I've seen issues with strange C headers interacting with certain rules in the past.

Made up example:

```
= %%%%%%%%%%%%%%%%%%%%%%%%%%%%%
= Header
= %%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

I've seen weird stuff like this... I think we may need:

- `%%` relevance: 0
- `%...%` 

---

You know... just make the relevance 0 period... I'm not sure that `%` operators are worthy of special relevance... it's used as modulus in many languages so % not % % as a pattens is rife for false positives with simple math formulas.  I'm going to say that this having relevance in the past has likely been a mistake.

Reply:
Ah, shame, I was hoping that we could precisely use `%…%` operators for a nice relevance boost since it’s so specific to R, but of course this overlooked its use as a comment fence.

I think rather than do any more whacky matching I’ll go with your last suggestion and just dial its relevance down to 0 everywhere.

Reply:
… hmm, nope, that leads to a failing detect test, the sample now gets detected as Vala. Which is weird, since that’s actually a *very* different language (looks a lot like Java), but hl.js has constant issues telling the two apart.

Reply:
>  but of course this overlooked its use as a comment fence.

And it's use as modulus...  if there were a super common R pattern perhaps there is still room for that:

```
% IDENT %
```

But I don't think just % anything % later works (and was far too overly broad before)...

Reply:
Why does Vala have such a high boost for preprocessor comments?

https://github.com/highlightjs/highlight.js/blob/6e182e0f306114d68dab261a63404ab50e40cb87/src/languages/vala.js#L51-L56

That’s completely unjustified, it’s an extremely common syntax for comments. Removing that relevance boost fixes the detection issue.

Reply:
My issues with `%...%` still stand (they aren't resolved by a change to Vala), but yes I agree that's Vala being misbehaved and would welcome a PR to fix that.  The default relevance of 1 should apply there.

Reply:
Hahaha, I guess we can include it here, please mention it in changelog though. :-)

Reply:
> My issues with `%...%` still stand

Are you sure ? It now has `relevance: 0` everywhere.

Reply:
No no  I meant that just changing Vala wouldn't solve the problem if that's what you were saying but I don't think it was. :-) So we're good.